### ISSUE TITLE
[C++] Problem parsing prototypes containing floating-point default values

### ISSUE THREAD

**Issue by bluescarni** [sphinx-doc/sphinx#5636]
Hello,

I am using sphinx + breathe to document a C++ software project. I have some function declarations which contain default floating-point parameters, which sometimes are expressed as literals in scientific notation, e.g., ``1e-6``. It seems like sphinx does not like them:

```
/home/yardbird/repos/pagmo2/doc/sphinx/docs/cpp/algorithms/de1220.rst:4: WARNING: Error when parsing function declaration.
If the function has no return type:
  Error in declarator or parameters and qualifiers
  Invalid definition: Expecting "," or ")" in parameters_and_qualifiers, got "e". [error at 183]
    pagmo::de1220::de1220(unsigned int gen = 1u, std::vector< unsigned int > allowed_variants = de1220_statics < void >::allowed_variants, unsigned int variant_adptv = 1u, double ftol = 1e-6, double xtol = 1e-6, bool memory = false, unsigned int seed = pagmo::random_device::next ())
    ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------^
```

I can provide a more detailed example, if needed.

**Comment by jakobandersen** [comment]
Thanks for reporting. It's a simple case of me not implementing all cases initially. You, or anyone else, are most welcome to make a PR. Essentially the regex at https://github.com/sphinx-doc/sphinx/blob/master/sphinx/domains/cpp.py#L297 should be extended, and some extra test combinations should be generated at around https://github.com/sphinx-doc/sphinx/blob/master/tests/test_domain_cpp.py#L127.

**Comment by jakobandersen** [comment]
It should be fixed in the 1.8 branch now.
